driver:temperature: Added support for max31865pmb1

jsanbuen opened the pull request:
Added driver implementation for max31865pmb1

More information: https://www.analog.com/media/en/technical-documentation/data-sheets/max31865.pdf

Signed-off-by: Jose Ramon San Buenaventura <jose.sanbuenaventura@analog.com>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/analogdevicesinc/no-OS?pullRequest=1578) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/analogdevicesinc/no-OS?pullRequest=1578) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/analogdevicesinc/no-OS?pullRequest=1578) it.</sub>

amiclaus commented on drivers/temperature/max31865pmb1/max31865.c:
```diff
@@ -0,0 +1,361 @@
+ */
+uint8_t max31865_read_fault(struct max31865_dev *device)
+{
+	uint8_t faultReg = MAX31865_FAULTSTAT_REG;
```
please drop camelcase `faultReg` -> `fault_reg`. this is valid for all the other variables throughout the code.

amiclaus commented on drivers/temperature/max31865pmb1/max31865.c:
```diff
@@ -0,0 +1,361 @@
+ * @param device MAX31865 descriptor
+ * @return The raw unsigned 8-bit FAULT status register
+ */
+uint8_t max31865_read_fault(struct max31865_dev *device)
```
Is this function really required? You can use the generic read function inline when necessary

amiclaus commented on drivers/temperature/max31865pmb1/max31865.c:
```diff
@@ -0,0 +1,361 @@
+ * @param val - register value
+ * @return 0 in case of success, negative error code otherwise
+ */
+int max31865_read_raw(struct max31865_dev *device, uint8_t *val)
```
please use standard function prototype for the read/write functions. see an example:
https://github.com/analogdevicesinc/no-OS/blob/d6ca269bc4b06260c18a1a767f00678a284f57a5/drivers/adc/adaq8092/adaq8092.h#L231-L236

amiclaus commented on drivers/temperature/max31865pmb1/max31865.c:
```diff
@@ -0,0 +1,361 @@
+ */
+int max31865_read_raw(struct max31865_dev *device, uint8_t *val)
+{
+	if((*val & 0x7F) != *val)
```
add please relevand macro definitions for these values: `0x7F`, `0x07`

amiclaus commented on drivers/temperature/max31865pmb1/max31865.c:
```diff
@@ -0,0 +1,361 @@
+	if(*val < 0x00 || *val > 0x07)
+		return ENXIO;
+
+	int ret;
```
variable declarations should be at the start of the function.

amiclaus commented on drivers/temperature/max31865pmb1/max31865.c:
```diff
@@ -0,0 +1,361 @@
+		*val &= 0x7F;
+
+	if(*val < 0x00 || *val > 0x07)
+		return ENXIO;
```
`-ENXIO`

amiclaus commented on drivers/temperature/max31865pmb1/max31865.c:
```diff
@@ -0,0 +1,361 @@
+int max31865_write_raw(struct max31865_dev *device, uint8_t val, uint8_t val2)
+{
+	if((val & 0x80) != 0x80)
+		val |= 0x80;
```
please add relevant macro definition for `0x80`

amiclaus commented on drivers/temperature/max31865pmb1/max31865.c:
```diff
@@ -0,0 +1,361 @@
+	if(val < 0x00 || val > 0x07)
+		return ENXIO;
+
+	int ret;
```
ditto: variable declaration should be done at the start of the function.

amiclaus commented on drivers/temperature/max31865pmb1/max31865.c:
```diff
@@ -0,0 +1,361 @@
+	faultReg &= ~0x2C;
+	faultReg |= 0x02;
+	uint8_t clrReg = 0x00;
+	max31865_write_raw(device, clrReg, faultReg);
```
return codes should be checked here for both read and write functions, therefore: `void max31865_clear_fault(struct max31865_dev *device)` -> `int max31865_clear_fault(struct max31865_dev *device)`

amiclaus commented on drivers/temperature/max31865pmb1/max31865.c:
```diff
@@ -0,0 +1,361 @@
+ * @param device MAX31865 descriptor
+ * @param b If true bias is enabled, else disabled
+ */
+void max31865_enable_bias(struct max31865_dev *device, bool b)
```
same here, check the return codes for the read/write functions and change the function type to `int`. Also i'f use a better naming for variable `b` -> `bias_en`

amiclaus commented on drivers/temperature/max31865pmb1/max31865.c:
```diff
@@ -0,0 +1,361 @@
+ * @param device MAX31865 descriptor
+ * @param b If true, auto conversion is enabled
+ */
+void max31865_auto_convert(struct max31865_dev *device, bool b)
```
ditto: change type to `int` and check the error codes.

amiclaus commented on drivers/temperature/max31865pmb1/max31865.c:
```diff
@@ -0,0 +1,361 @@
+ * @param device MAX31865 descriptor
+ * @param b If true, 50Hz noise is filtered, else 60Hz(default)
+ */
+void max31865_enable_50Hz(struct max31865_dev *device, bool b)
```
ditto: change type to `int` and check the error codes.

amiclaus commented on drivers/temperature/max31865pmb1/max31865.c:
```diff
@@ -0,0 +1,361 @@
+ * @param lower raw lower threshold
+ * @param upper raw upper threshold
+ */
+void max31865_set_threshold(struct max31865_dev *device, uint16_t lower,
```
ditto: change type to `int` and check the error codes.

amiclaus commented on drivers/temperature/max31865pmb1/max31865.c:
```diff
@@ -0,0 +1,361 @@
+	uint16_t lsb;
+	uint16_t msb;
+
+	lsb = upper & 0xFF;
```
you can use here `no_os_field_get`

amiclaus commented on drivers/temperature/max31865pmb1/max31865.c:
```diff
@@ -0,0 +1,361 @@
+	max31865_write_raw(device, MAX31865_HFAULTLSB_REG, (uint8_t)msb);
+
+	lsb = lower & 0xFF;
+	msb = lower  >> 8;
```
ditto: use `no_os_field_get`

amiclaus commented on drivers/temperature/max31865pmb1/max31865.c:
```diff
@@ -0,0 +1,361 @@
+ * @param device MAX31865 descriptor
+ * @return The raw unsigned 16-bit value, NOT temperature value
+ */
+uint16_t max31865_get_lower_threshold(struct max31865_dev *device)
```
use a pointer variable in the function prototype to store the threshold. return afferent error code from the read/write functions.

amiclaus commented on drivers/temperature/max31865pmb1/max31865.c:
```diff
@@ -0,0 +1,361 @@
+ * @param device MAX31865 descriptor
+ * @return The raw unsigned 16-bit value, NOT temperature value
+ */
+uint16_t max31865_get_upper_threshold(struct max31865_dev *device)
```
ditto: use a pointer variable in the function prototype to store the threshold. return afferent error code from the read/write functions.

amiclaus commented on drivers/temperature/max31865pmb1/max31865.c:
```diff
@@ -0,0 +1,361 @@
+ * @param device MAX31865 descriptor
+ * @param wires The number of wires in enum format
+ */
+void max31865_set_wires(struct max31865_dev *device, max31865_numwires_t wires)
```
`int` type should be used and check the error codes inside the function.

amiclaus commented on drivers/temperature/max31865pmb1/max31865.h:
```diff
@@ -0,0 +1,143 @@
+#define MAX31865_FAULT_OVUV 0x04
+
+
+typedef enum max31865_numwires {
```
remove typedefs please

amiclaus commented on drivers/temperature/max31865pmb1/max31865.c:
```diff
@@ -0,0 +1,361 @@
+ * @param device MAX31865 descriptor
+ * @return The raw unsigned 16-bit value, NOT temperature
+ */
+uint16_t max31865_read_RTD(struct max31865_dev *device)
```
use `int` and a pointer variable to store the RTD_REG value.

amiclaus reviewed (commented):
some comments on my side, please sign the cla and fix also the code style issue, see the guide:
https://github.com/analogdevicesinc/no-OS/wiki/Code-Style-guidelines

jsanbuen commented:
Referring to MAX31865's datasheet, it communicates to master via half duplex SPI (TX first then RX). Upon checking "maxim_spi.c", the transaction used defaults to the full duplex SPI (simultaneous TX and RX based on SPI_WIDTH_STANDARD). Will also do experiments with example codes etc. to find necessary changes needed for communicating properly with the MAX31865.

amiclaus commented on drivers/temperature/max31865pmb1/max31865.c:
```diff
@@ -0,0 +1,441 @@
+		return -ENXIO;
+
+	ret = no_os_spi_transfer(device->comm_desc, &temp_xfer, 1);
+
```
nit: remove extra line

amiclaus commented on drivers/temperature/max31865pmb1/max31865.c:
```diff
@@ -0,0 +1,441 @@
+
+	*reg_data = raw_array[1];
+
+	return ret;
```
return 0;

amiclaus commented on drivers/temperature/max31865pmb1/max31865.c:
```diff
@@ -0,0 +1,441 @@
+	if (ret)
+		return ret;
+
+	return ret;
```
return 0;

amiclaus commented on drivers/temperature/max31865pmb1/max31865.c:
```diff
@@ -0,0 +1,441 @@
+	reg_data &= ~0x2C;
+	reg_data |= MAX31865_CONFIG_FAULTSTAT;
+
+	ret = max31865_write(device, MAX31865_CONFIG_REG, reg_data);
```
return directly

amiclaus commented on drivers/temperature/max31865pmb1/max31865.c:
```diff
@@ -0,0 +1,441 @@
+	if(ret)
+		return ret;
+
+	reg_data &= ~0x2C;
```
please provide relevant macro for 0x2C. is it a mask?

amiclaus commented on drivers/temperature/max31865pmb1/max31865.c:
```diff
@@ -0,0 +1,441 @@
+	else
+		reg_data &= ~MAX31865_CONFIG_BIAS;
+
+	ret = max31865_write(device, MAX31865_CONFIG_REG, reg_data);
```
return directly

amiclaus commented on drivers/temperature/max31865pmb1/max31865.c:
```diff
@@ -0,0 +1,441 @@
+	else
+		reg_data &= ~MAX31865_CONFIG_FILT50HZ;
+
+	ret = max31865_write(device, MAX31865_CONFIG_REG, reg_data);
```
return directly

amiclaus commented on drivers/temperature/max31865pmb1/max31865.c:
```diff
@@ -0,0 +1,441 @@
+	if(ret)
+		return ret;
+
+	ret = max31865_write(device, MAX31865_LFAULTLSB_REG, (uint8_t)lsb);
```
return directly.

amiclaus commented on drivers/temperature/max31865pmb1/max31865.h:
```diff
@@ -0,0 +1,145 @@
+	struct no_os_spi_init_param spi_init;
+};
+
+
```
nit: extra line.

amiclaus commented on drivers/temperature/max31865pmb1/max31865.h:
```diff
@@ -0,0 +1,145 @@
+/********************** Macros and Constants Definitions **********************/
+/******************************************************************************/
+
+#define MAX31865_READ_MASK  0x7F
```
remove extra space.

amiclaus reviewed (commented):
some comments on my side. I see you did recently a force push with some changes related to the previous review. Please provide also a changelog each time you push some modifications.

jsanbuen commented:
Changelogs:

-removed additional spaces based on previous comments
-changed return values to either "0" or "directly" depending on function and comments
-added macro define for ~0x2C / 0xD3 as MAX31865_CONFIG_CLRFAULT_MASK
-removed other macro defines that aren't used.

CiprianRegus commented on drivers/temperature/max31865pmb1/max31865.c:
```diff
@@ -0,0 +1,424 @@
+	if(ret)
+		return ret;
+
+	*rtd_reg = ((uint16_t)reg_data << 8);
```
The cast is not required

CiprianRegus commented on drivers/temperature/max31865pmb1/max31865.c:
```diff
@@ -0,0 +1,424 @@
+ */
+int max31865_set_wires(struct max31865_dev *device, bool is_odd_wire)
+{
+	uint8_t reg_data = 0x00;
```
you don't have to initialize this

CiprianRegus commented on drivers/temperature/max31865pmb1/max31865.c:
```diff
@@ -0,0 +1,424 @@
+	int ret;
+
+	ret = max31865_read(device, MAX31865_CONFIG_REG, &reg_data);
+
```
Remove the empty line.

CiprianRegus commented on drivers/temperature/max31865pmb1/max31865.c:
```diff
@@ -0,0 +1,424 @@
+	if(ret)
+		return ret;
+
+	*low_threshold = (((uint16_t)low_msb << 8) | (uint16_t)low_lsb);
```
The casts are not needed.

CiprianRegus commented on drivers/temperature/max31865pmb1/max31865.c:
```diff
@@ -0,0 +1,424 @@
+	if(ret)
+		return ret;
+
+	*up_threshold = (((uint16_t)high_msb << 8) | (uint16_t)high_lsb);
```
Same here.

CiprianRegus commented on drivers/temperature/max31865pmb1/max31865.c:
```diff
@@ -0,0 +1,424 @@
+	if(ret)
+		return ret;
+
+	ret = max31865_read(device, MAX31865_CONFIG_REG, &reg_data);
+	if(ret)
+		return ret;
+
+	reg_data |= MAX31865_CONFIG_1SHOT;
+
+	ret = max31865_write(device, MAX31865_CONFIG_REG, reg_data);
+	if(ret)
+		return ret;
+
```
You could improve this by implementing an `reg_update()` function. There are a few other places where this would be the case.

jsanbuen replied:
Hi Ciprian,
Does this mean that I need to implement a "max31865_reg_update()" function?

CiprianRegus commented on drivers/temperature/max31865pmb1/max31865.c:
```diff
@@ -0,0 +1,424 @@
+	if (!descriptor)
+		return -ENOMEM;
+
+	ret = no_os_spi_init(&descriptor->comm_desc, &init_param->spi_init);
+	if (ret) {
+		free(descriptor);
+		return ret;
+	}
```
This is not wrong, but the way we usually implement these checks like I did here: https://github.com/analogdevicesinc/no-OS/blob/master/drivers/adc-dac/ad74413r/ad74413r.c#L999-L1042

CiprianRegus reviewed (commented):
Looking pretty good, just a few minor remarks.

jsanbuen commented:
Changelog/s:
-added reg_update function and implemented on appropriate lines.
-removed extra spaces.
-removed cast on suggested lines.
-revised error checking on init function based on suggested similar device driver

buha commented:
@jsanbuen can you please get rid of the "Merged" commit ? it should be just one commit with your change. Use Git's rebase feature rather than merge to make sure your branch is on top master.
